From d1593016918a0a784311ff7f253d88499175d406 Mon Sep 17 00:00:00 2001
From: Garrett Brown <themagnificentmrb@gmail.com>
Date: Tue, 13 Oct 2026 02:31:49 -0700
Subject: [PATCH] [cmake] Fix Windows LCMS2 definitions after rediscovery

Restored dependencies use pkg-config, bypassing the internal-build
CMS_DLL definition and breaking x86 links. Detect DLL imports when
rediscovering LCMS2 so shared consumers retain CMS_DLL while static
libraries remain usable.
---
 cmake/modules/FindLCMS2.cmake | 19 +++++++++++++++----
 1 file changed, 15 insertions(+), 4 deletions(-)

diff --git a/cmake/modules/FindLCMS2.cmake b/cmake/modules/FindLCMS2.cmake
index 5968d9a..ee40759 100644
--- a/cmake/modules/FindLCMS2.cmake
+++ b/cmake/modules/FindLCMS2.cmake
@@ -81,15 +81,22 @@ if(NOT TARGET ${APP_NAME_LC}::${CMAKE_FIND_PACKAGE_NAME})
     if(TARGET PkgConfig::${${CMAKE_FIND_PACKAGE_NAME}_SEARCH_NAME_PC} AND NOT TARGET ${${${CMAKE_FIND_PACKAGE_NAME}_MODULE}_BUILD_NAME})
       add_library(${APP_NAME_LC}::${CMAKE_FIND_PACKAGE_NAME} ALIAS PkgConfig::${${CMAKE_FIND_PACKAGE_NAME}_SEARCH_NAME_PC})
       add_library(LIBRARY::${CMAKE_FIND_PACKAGE_NAME} ALIAS PkgConfig::${${CMAKE_FIND_PACKAGE_NAME}_SEARCH_NAME_PC})
+
+      if(WIN32 OR WINDOWS_STORE)
+        include(CheckSymbolExists)
+        include(CMakePushCheckState)
+        cmake_push_check_state(RESET)
+        set(CMAKE_REQUIRED_DEFINITIONS -DCMS_DLL)
+        set(CMAKE_REQUIRED_LIBRARIES ${APP_NAME_LC}::${CMAKE_FIND_PACKAGE_NAME})
+        unset(${${CMAKE_FIND_PACKAGE_NAME}_MODULE}_SHARED_LIB CACHE)
+        check_symbol_exists(cmsCreateTransform lcms2.h ${${CMAKE_FIND_PACKAGE_NAME}_MODULE}_SHARED_LIB)
+        cmake_pop_check_state()
+      endif()
     elseif(TARGET ${${${CMAKE_FIND_PACKAGE_NAME}_MODULE}_BUILD_NAME})
       SETUP_BUILD_TARGET()
 
       add_dependencies(${APP_NAME_LC}::${CMAKE_FIND_PACKAGE_NAME} ${${${CMAKE_FIND_PACKAGE_NAME}_MODULE}_BUILD_NAME})
 
-      if(WIN32 OR WINDOWS_STORE)
-        list(APPEND ${${CMAKE_FIND_PACKAGE_NAME}_MODULE}_COMPILE_DEFINITIONS CMS_DLL)
-      endif()
-
       # We are building as a requirement, so set LIB_BUILD property to allow calling
       # modules to know we will be building, and they will want to rebuild as well.
       # Property must be set on actual TARGET and not the ALIAS
@@ -98,6 +105,10 @@ if(NOT TARGET ${APP_NAME_LC}::${CMAKE_FIND_PACKAGE_NAME})
       add_library(LIBRARY::${CMAKE_FIND_PACKAGE_NAME} ALIAS ${APP_NAME_LC}::${CMAKE_FIND_PACKAGE_NAME})
     endif()
 
+    if((WIN32 OR WINDOWS_STORE) AND ${${CMAKE_FIND_PACKAGE_NAME}_MODULE}_SHARED_LIB)
+      list(APPEND ${${CMAKE_FIND_PACKAGE_NAME}_MODULE}_COMPILE_DEFINITIONS CMS_DLL)
+    endif()
+
     list(APPEND ${${CMAKE_FIND_PACKAGE_NAME}_MODULE}_COMPILE_DEFINITIONS HAVE_LCMS2
                                                                          CMS_NO_REGISTER_KEYWORD)
 
-- 
2.53.0

